crates/holochain_cascade/src/authority/get_agent_activity_query/must_get_agent_activity.rs

```diff
@@ -83,32 +83,57 @@ pub fn filter_then_check(
     }
 }
 
-/// Find the filters sequence bounds.
+/// Find the filter's sequence bounds.
 fn find_bounds(
     txn: &Transaction,
     scratch: Option<&Scratch>,
     author: &AgentPubKey,
     filter: ChainFilter,
 ) -> StateQueryResult<Sequences> {
-    let mut statement = txn.prepare(ACTION_HASH_TO_SEQ)?;
+    let mut by_hash_statement = txn.prepare(ACTION_HASH_TO_SEQ)?;
 
     // Map an action hash to its sequence.
-    let get_seq = move |hash: &ActionHash| {
+    let get_seq_from_hash = |hash: &ActionHash| {
         if let Some(scratch) = scratch {
             if let Some(action) = scratch.actions().find(|a| a.action_address() == hash) {
                 return Ok(Some(action.action().action_seq()));
             }
         }
-        let result = statement
+        let result = by_hash_statement
                 .query_row(named_params! {":hash": hash, ":author": author, ":activity": ChainOpType::RegisterAgentActivity}, |row| {
                     row.get(0)
                 })
                 .optional()?;
         Ok(result)
     };
 
+    // Get the oldest action in a given time period.
+    let mut by_ts_statement = txn.prepare(TS_TO_SEQ)?;
+    let get_seq_from_ts = |until: Timestamp| {
+        // Search in DB
+        let db_result = by_ts_statement
+          .query_row(named_params! {":until": until.as_micros(), ":author": author, ":activity": ChainOpType::RegisterAgentActivity}, |row| {
+              row.get(0)
+          })
+          .optional()?;
+        // Search in scratch
+        let scratch_result = scratch.and_then(|s| {
+            s.actions()
+                .filter_map(|a| {
```

Comment:
It's unclear whether we can rely on the ordering of actions in the scratch to be by sequence number. I think that is the case but it's not documented to be a requirement. 

With the actions in order, this is actually a `find_map` without the `min` and not as much needs to be iterated but this code is resilient to problem behavior in the scratch space so perhaps this is best!